File: erpnext/hr/report/employee_leave_balance/employee_leave_balance.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["Employee Leave Balance"] = {
	"filters": [
                {
                        "fieldname":"leave_type",
                        "label": __("Leave Type"),
                        "fieldtype": "Link",
                        "options": "Leave Type",
			"reqd": 1,
                },
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.year_start()
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.year_end()
		},
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"reqd": 1,
			"default": frappe.defaults.get_user_default("Company")
		},
		{
                        "fieldname":"branch",
                        "label": __("Branch"),
                        "fieldtype": "Link",
                        "options": "Branch",
                },
		{
                        "fieldtype": "Break"
                },
		{
			"fieldname": "employee_type",
			"label": __("Employee Type"),
			"fieldtype": "Link",
			"options": "Employment Type" 
		},
                {
                        "fieldname":"employee",
                        "label": __("Employee"),
                        "fieldtype": "Link",
                        "options": "Employee",
			"get_query": function() {
				var branch = frappe.query_report.filters_by_name.branch.get_value();
				var employee_type = frappe.query_report.filters_by_name.employee_type.get_value();
				if(branch) {
					return {"doctype": "Employee", "filters": {"branch": branch, "status": "Active"}}
				}
				else if (employee_type){
				return {"doctype": "Employee", "filters": {"employment_type": employee_type, "status": "Active"}}
				} 
				else 
				{
					return {"doctype": "Employee", "filters": {"status": "Active"}}
				}
			}
                }
	]
}

File: erpnext/hr/report/employee_leave_balance/employee_leave_balance.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from erpnext.hr.doctype.leave_application.leave_application \
	import get_leave_allocation_records, get_leave_balance_on, get_approved_leaves_for_period


def execute(filters=None):
	if filters.leave_type:
		leave_types = frappe.db.sql_list("select name from `tabLeave Type` where name = %s order by name asc", filters.leave_type)
	else:
		leave_types = frappe.db.sql_list("select name from `tabLeave Type` order by name asc")

	columns = get_columns(leave_types)
	data = get_data(filters, leave_types)

	return columns, data

def get_columns(leave_types):
	columns = [
		_("Employee") + ":Link/Employee:120",
		_("Employee Name") + "::150",
		_("Department") +"::150",
		_("Branch") +"::160",

	]

	for leave_type in leave_types:
		columns.append(_(leave_type) + " " + _("Taken") + ":Float:160")
		columns.append(_(leave_type) + " " + _("Balance") + ":Float:160")

	return columns

def get_data(filters, leave_types):

	allocation_records_based_on_to_date = get_leave_allocation_records(filters.to_date)
	filters_dict = { "status": "Active", "company": filters.company}

        if filters.branch:
                filters_dict['branch'] = filters.branch
        if filters.employee:
                filters_dict['name'] = filters.employee

	if filters.employee_type:
		filters_dict['employment_type'] = filters.employee_type

	active_employees = frappe.get_all("Employee",
		filters = filters_dict,
		fields = ["name", "employee_name", "department", "branch"])

	data = []
	for employee in active_employees:
		row = [employee.name, employee.employee_name, employee.department, employee.branch]

		for leave_type in leave_types:
			# leaves taken
			leaves_taken = get_approved_leaves_for_period(employee.name, leave_type,
				filters.from_date, filters.to_date)
			
			# closing balance
			closing = get_leave_balance_on(employee.name, leave_type, filters.to_date,
				allocation_records_based_on_to_date.get(employee.name, frappe._dict()))
			
			row += [leaves_taken, closing]

		data.append(row)

	return data
